test/dpe_verification/transport.go

```diff
@@ -0,0 +1,239 @@
+		log.Fatal(err)
+	}
+
+	cRom := C.CBytes(rom)
+	defer C.free(cRom)
+
+	params.rom.data = (*C.uchar)(cRom)
+	params.rom.len = C.uintptr_t(len(rom))
+
+	status := C.caliptra_model_init_default(params, &CALIPTRA_C_MODEL)
+	if status != 0 {
+		panic("Failed to initialize caliptra model")
+	}
+
+	return CALIPTRA_C_MODEL
+}
+
+// Start the Emulator.
+func (s *CptraModel) PowerOn() error {
+	model := getHWModel()
+	if model == nil {
+		panic("Failed to get HW model")
+	}
+
+	C.caliptra_bootfsm_go()
+
+	if C.set_fuses() != 0 {
+		panic("Failed to set fuses")
+	}
+
+	C.caliptra_ready_for_firmware()
+
+	bundlePath := os.Getenv("FW_PATH")
+	fwBundle, err := ioutil.ReadFile(bundlePath)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var bundleBuf C.struct_caliptra_buffer
+
+	cBundle := C.CBytes(fwBundle)
+	defer C.free(cBundle)
+
+	bundleBuf.data = (*C.uchar)(cBundle)
+	bundleBuf.len = C.uintptr_t(len(fwBundle))
+	C.caliptra_upload_fw(&bundleBuf, false)
+
+	return nil
+}
+
+// Kill the emulator in a way that it can cleanup before closing.
+func (s *CptraModel) PowerOff() error {
+	C.caliptra_model_destroy(CALIPTRA_C_MODEL)
+	CALIPTRA_C_MODEL = nil
+
+	return nil
+}
+
+func (s *CptraModel) SendCmd(buf []byte) ([]byte, error) {
+	var req C.struct_caliptra_invoke_dpe_req
+	var resp C.struct_caliptra_invoke_dpe_resp
+
+	// Caliptra expects all DPE commands to fill the whole data buffer
+	C.memcpy(unsafe.Pointer(&req.data), unsafe.Pointer(&buf[0]), C.size_t(len(buf)))
+	req.data_size = C.uint32_t(512)
+
+	cptraStatus := C.caliptra_invoke_dpe_command(&req, &resp, false)
+	if cptraStatus != 0 {
+		return []byte{}, fmt.Errorf("Failed to send DPE command, error 0x%08x", int(cptraStatus))
+	}
+
+	mboxStatus := C.caliptra_read_fw_non_fatal_error()
+	if mboxStatus != 0 {
+		return []byte{}, fmt.Errorf("Caliptra mailbox returned error 0x%08x", int(mboxStatus))
+	}
+
+	respPtr := &resp
+
+	// Calculate the offset to the union member based on data_size
+	offset := unsafe.Sizeof(resp.cpl) + unsafe.Sizeof(resp.data_size)
+	var selectedBytes []byte
+
+	selectedBytes = C.GoBytes(unsafe.Pointer(uintptr(unsafe.Pointer(respPtr))+offset), (C.int)(resp.data_size))
+
+	return selectedBytes, nil
+}
+
+func (s *CptraModel) GetSupport() *verification.Support {
+	// Expected support vector
+	return &verification.Support{
+		Simulation:    true,
+		ExtendTci:     true,
+		AutoInit:      true,
+		Tagging:       true,
+		RotateContext: true,
+		X509:          true,
+		Csr:           true,
+		IsSymmetric:   true,
+		InternalInfo:  true,
+		InternalDice:  true,
+		IsCA:          true,
+	}
+}
+
+func (s *CptraModel) GetSupportedLocalities() []uint32 {
+	// TODO: Make this configurable, since it will be SoC specific
+	return []uint32{0, 1}
+}
+
+func (s *CptraModel) SetLocality(locality uint32) {
+	// TODO: Hook this up to HwModel
+	s.currentLocality = locality
+}
+
+func (s *CptraModel) GetLocality() uint32 {
+	return s.currentLocality
+}
+
+func (s *CptraModel) GetMaxTciNodes() uint32 {
```

Comment:
Not too familiar with go/rust integration, so not sure if this is possible but is there a way we can import these constants from dpe/src/lib.rs?

Reply:
Technically yes

1. In build.rs, generate C headers with cbindgen
2. Include these headers in cgo
3. Use the constants with C.*

I'll leave this for now and we can maybe look into doing this in the future.